features/filesystem/bd/BufferedBlockDevice.cpp

```diff
@@ -0,0 +1,229 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "BufferedBlockDevice.h"
+#include "mbed_assert.h"
+#include <algorithm>
+#include <string.h>
+
+static inline uint32_t align_down(bd_size_t val, bd_size_t size)
+{
+    return val / size * size;
+}
+
+BufferedBlockDevice::BufferedBlockDevice(BlockDevice *bd)
+    : _bd(bd), _bd_program_size(0), _curr_aligned_addr(0), _flushed(true), _cache(0)
+{
+}
+
+BufferedBlockDevice::~BufferedBlockDevice()
+{
+    deinit();
+    if (_cache) {
+        delete[] _cache;
+    }
+}
+
+int BufferedBlockDevice::init()
+{
+    int err = _bd->init();
+    if (err) {
+        return err;
+    }
+
+    _bd_program_size = _bd->get_program_size();
+
+    if (!_cache) {
+        _cache = new uint8_t[_bd_program_size];
+    }
+
+    _curr_aligned_addr = _bd->size();
+    _flushed = true;
+
+    return 0;
+}
+
+int BufferedBlockDevice::deinit()
+{
+    return _bd->deinit();
+}
+
+int BufferedBlockDevice::flush()
+{
+    if (!_flushed) {
+        int ret = _bd->program((const void *) _cache, _curr_aligned_addr, _bd_program_size);
+        if (ret) {
+            return ret;
+        }
+        _flushed = true;
+    }
+    return 0;
+}
+
+int BufferedBlockDevice::sync()
+{
+    int ret = flush();
+    if (ret) {
+        return ret;
+    }
+    return _bd->sync();
+}
+
+int BufferedBlockDevice::read(void *b, bd_addr_t addr, bd_size_t size)
+{
+    MBED_ASSERT(_cache);
+    bool moved_unit = false;
+
+    bd_addr_t aligned_addr = align_down(addr, _bd_program_size);
+
+    uint8_t *buf = (uint8_t *) b;
+
+    if (aligned_addr != _curr_aligned_addr) {
+        // Need to flush if moved to another program unit
+        flush();
+        _curr_aligned_addr = aligned_addr;
+        moved_unit = true;
+    }
+
+    while (size) {
+        _curr_aligned_addr = align_down(addr, _bd_program_size);
+        if (moved_unit) {
+            int ret = _bd->read(_cache, _curr_aligned_addr, _bd_program_size);
+            if (ret) {
+                return ret;
+            }
+        }
+        bd_addr_t offs_in_buf = addr - _curr_aligned_addr;
+        bd_size_t chunk = std::min(_bd_program_size - offs_in_buf, size);
+        memcpy(buf, (const void *)(_cache + offs_in_buf), chunk);
+        moved_unit = true;
+        buf += chunk;
+        addr += chunk;
+        size -= chunk;
+    }
+
+    return 0;
+}
+
+int BufferedBlockDevice::program(const void *b, bd_addr_t addr, bd_size_t size)
+{
+    MBED_ASSERT(_cache);
+    int ret;
+    bool moved_unit = false;
+
+    bd_addr_t aligned_addr = align_down(addr, _bd_program_size);
+
+    uint8_t *buf = (uint8_t *) b;
```

Comment:
Should be `static_cast<>` to check `const` and assigning to `const uint8_t *`.

Reply:
Correct. Fixed.